src/LinAlg/hiopLinSolverSparsePARDISO.cpp

```diff
@@ -0,0 +1,523 @@
+
+    iparm_[2] = num_threads_;
+    iparm_[1] = 2;  // 2 is for metis, 0 for min degree
+    iparm_[7] = 3;  // # iterative refinements
+    iparm_[10] = 1; // scaling for IPM KKT; used with IPARM(13)=1 or 2
+    iparm_[12] = 2; // improved accuracy for IPM KKT; used with IPARM(11)=1; 
+                    // if needed, use 2 for advanced matchings and higer accuracy.
+    iparm_[23] = 1;  // Parallel Numerical Factorization 
+                    // (0=used in the last years, 1=two-level scheduling)
+
+    /* symbolic analysis from PARDISO */
+    int phase = 11; //analysis
+    int nrhs = 1;
+  
+    pardiso_d(pt_ , &maxfct_, &mnum_, &mtype_, &phase,
+              &n_, kVal_, kRowPtr_, jCol_,
+              NULL, &nrhs, 
+              iparm_, &msglvl_, NULL, NULL, &error_, dparm_);
+    if ( error_ != 0) {
+      printf ("PardisoSolver - ERROR during symbolic factorization: %d\n", error_ );
+      assert(false);
+    }
+  
+  }
+
+  int hiopLinSolverIndefSparsePARDISO::matrixChanged()
+  {
+    assert(n_==M.n() && M.n()==M.m());
+    assert(n_>0);
+
+    nlp_->runStats.linsolv.tmFactTime.start();
+
+    if(!is_initialized_) {
+      this->firstCall();
+      is_initialized_ = true;
+    } else {
+      // update matrix
+      int rowID_tmp{0};
+      for(int k=0;k<nnz_;k++){
+        kVal_[k] = M.M()[index_covert_CSR2Triplet_[k]];
+      }
+      for(int i=0;i<n_;i++){
+        if(index_covert_extra_Diag2CSR_[i] != -1)
+          kVal_[index_covert_extra_Diag2CSR_[i]] += M.M()[M.numberOfNonzeros()-n_+i];
+      }
+    }
+
+    /* do numerical factorization */
+    int phase = 22;
+    int nrhs = 1;
+  
+    pardiso_d(pt_ , &maxfct_, &mnum_, &mtype_, &phase,
+              &n_, kVal_, kRowPtr_, jCol_,
+              NULL, &nrhs, 
+              iparm_, &msglvl_, NULL, NULL, &error_, dparm_);
+
+    if ( error_ != 0) {
+      printf ("PardisoSolver - ERROR during numerical factorization: %d\n", error_ );
+      assert(false);
+    }
+
+    nlp_->runStats.linsolv.tmInertiaComp.start();
+    
+    int negEigVal = iparm_[22];
+    if(iparm_[21]+iparm_[22] != n_) {
+      // singular matrix
+      negEigVal = -1;
+    }
+
+    nlp_->runStats.linsolv.tmInertiaComp.stop();
+
+    return negEigVal;
+  }
+
+  bool hiopLinSolverIndefSparsePARDISO::solve(hiopVector& b)
+  {
+    assert(n_==M.n() && M.n()==M.m());
+    assert(n_>0);
+    assert(b.get_size()==M.n());
+
+    nlp_->runStats.linsolv.tmTriuSolves.start();
+
+    /* do backsolve */
+    hiopVectorPar* x = dynamic_cast<hiopVectorPar*>(&b);
+    assert(x != nullptr);
+    if(rhs_==nullptr) {
+      rhs_ = dynamic_cast<hiopVectorPar*>(x->new_copy());
+    } else {
+      rhs_->copyFrom(*x);
+    }
+    double* dx = x->local_data();
+    double* drhs = rhs_->local_data();
+
+    int phase = 33;
+    int nrhs = 1;
+  
+    pardiso_d(pt_ , &maxfct_, &mnum_, &mtype_, &phase,
+              &n_, kVal_, kRowPtr_, jCol_,
+              NULL, &nrhs, 
+              iparm_, &msglvl_,
+              
```

Comment:
is this an (unnecessary) extra line?

Reply:
which line? 328 or 329? seems to me both are required.

Reply:
328 seems to be an empty line

Reply:
got it!